File: medi/src/app/core/services/order.service.ts
import { Injectable, signal, computed, inject } from '@angular/core';
import { StoreOrder, PrescribedOrderItem, OrderPatientInfo, OrderStatus, OrderPaymentMethod } from '../models/order.model';
import { User } from '../models/auth.model';
import { AuthService } from './auth.service';
import { ApiService } from './api.service';

export const INITIAL_ORDERS: StoreOrder[] = [];

@Injectable({
  providedIn: 'root'
})
export class OrderService {
  private readonly ORDERS_STORAGE_KEY = 'medi_store_orders_v1';
  private authService = inject(AuthService);
  private api = inject(ApiService);

  readonly orders = signal<StoreOrder[]>(this.loadOrders());

  constructor() {
    this.syncOrdersFromBackend();
  }

  syncOrdersFromBackend(): void {
    this.api.getOrders().subscribe({
      next: (backendOrders) => {
        if (backendOrders && backendOrders.length > 0) {
          const mapped: StoreOrder[] = backendOrders.map((o: any) => ({
            id: String(o.id || o.orderNumber),
            orderNumber: o.orderNumber,
            orderType: o.orderType,
            storeId: String(o.storeId || 'store-1'),
            storeName: o.storeName || 'MediCare Pharmacy',
            placedBy: {
              userId: String(o.userId || ''),
              userName: o.userName || '',
              userRole: o.userRole || 'CUSTOMER',
              userPhone: o.userPhone || '',
              doctorRegNo: o.doctorRegNo,
              doctorSpecialty: o.doctorSpecialty
            },
            patient: {
              patientName: o.patientName,
              patientAge: o.patientAge,
              patientGender: o.patientGender,
              patientPhone: o.patientPhone,
              diagnosis: o.diagnosis,
              vitalsBp: o.patientVitalsBp,
              vitalsPulse: o.patientVitalsPulse,
              vitalsWeight: o.patientVitalsWeight,
              vitalsTemp: o.patientVitalsTemp,
              vitalsSpo2: o.patientVitalsSpo2,
              familyMemberName: o.familyMemberName
            },
            familyMemberName: o.familyMemberName,
            patientVitalsBp: o.patientVitalsBp,
            patientVitalsPulse: o.patientVitalsPulse,
            patientVitalsWeight: o.patientVitalsWeight,
            patientVitalsTemp: o.patientVitalsTemp,
            patientVitalsSpo2: o.patientVitalsSpo2,
            items: (o.items || []).map((i: any) => ({
              medicineId: String(i.medicineId),
              medicineName: i.medicineName,
              genericName: i.genericName,
              packaging: i.packaging,
              quantity: Number(i.quantity),
              unitPrice: Number(i.unitPrice),
              total: Number(i.total),
              dosage: i.dosage,
              timing: i.timing,
              durationDays: i.durationDays
            })),
            prescriptionNotes: o.prescriptionNotes,
            deliveryAddress: o.deliveryAddress,
            prescriptionPhotoUrl: o.prescriptionPhotoUrl,
            paymentMethod: o.paymentMethod,
            paymentStatus: o.paymentStatus,
            orderStatus: o.orderStatus,
            totalAmount: Number(o.totalAmount),
            createdAt: o.createdAt || new Date().toISOString(),
            packedAt: o.packedAt,
            dispatchedAt: o.dispatchedAt,
            completedAt: o.completedAt,
            auditTrail: (o.auditTrail || []).map((a: any) => ({
              timestamp: a.timestamp,
              action: a.action,
              performedBy: a.performedBy
            }))
          }));
          this.orders.set(mapped);
          this.saveOrders(mapped);
        }
      },
      error: () => {}
    });
  }

  // Computeds for Store Owner
  readonly pendingOrders = computed(() => {
    return this.orders().filter(o => o.orderStatus === 'NEW_RECEIVED');
  });

  readonly packedOrders = computed(() => {
    return this.orders().filter(o => o.orderStatus === 'PACKED');
  });

  readonly codOrders = computed(() => {
    return this.orders().filter(o => o.paymentMethod === 'COD');
  });

  readonly doctorOrders = computed(() => {
    return this.orders().filter(o => o.orderType === 'DOCTOR_PRESCRIPTION' || o.orderType === 'DOCTOR_DRUG_ORDER');
  });

  private loadOrders(): StoreOrder[] {
    try {
      const saved = localStorage.getItem(this.ORDERS_STORAGE_KEY);
      if (saved) return JSON.parse(saved);
    } catch (e) {
      console.warn('Failed to load orders from storage', e);
    }
    return [];
  }

  private saveOrders(orders: StoreOrder[]): void {
    this.orders.set(orders);
    try {
      localStorage.setItem(this.ORDERS_STORAGE_KEY, JSON.stringify(orders));
    } catch (e) {
      console.error('Failed to save orders to storage', e);
    }
  }

  // 1. Doctor Prescription & Drug Order Creation
  createDoctorOrder(data: {
    storeId: string;
    storeName: string;
    doctorUser: User;
    patient: OrderPatientInfo;
    items: PrescribedOrderItem[];
    prescriptionNotes?: string;
    deliveryAddress?: string;
    paymentMethod: OrderPaymentMethod;
  }): StoreOrder {
    const totalAmount = data.items.reduce((sum, i) => sum + i.total, 0);
    const now = new Date();
    const orderNumber = 'ORD-DOC-' + String(this.orders().length + 1).padStart(4, '0');

    const newOrder: StoreOrder = {
      id: 'ord-' + Date.now(),
      orderNumber,
      orderType: 'DOCTOR_PRESCRIPTION',
      storeId: data.storeId,
      storeName: data.storeName,
      placedBy: {
        userId: data.doctorUser.id,
        userName: data.doctorUser.name,
        userRole: 'DOCTOR',
        userPhone: data.doctorUser.phone,
        doctorRegNo: data.doctorUser.doctorRegNo,
        doctorSpecialty: data.doctorUser.doctorSpecialty
      },
      patient: data.patient,
      items: data.items,
      prescriptionNotes: data.prescriptionNotes,
      deliveryAddress: data.deliveryAddress,
      paymentMethod: data.paymentMethod,
      paymentStatus: data.paymentMethod === 'ONLINE_PAID' ? 'PAID' : 'PENDING_COLLECTION',
      orderStatus: 'NEW_RECEIVED',
      totalAmount,
      createdAt: now.toISOString(),
      auditTrail: [
        {
          timestamp: now.toISOString(),
          action: 'Prescription Order Placed by Doctor',
          performedBy: `${data.doctorUser.name} (Reg: ${data.doctorUser.doctorRegNo})`
        }
      ]
    };

    const updated = [newOrder, ...this.orders()];
    this.saveOrders(updated);

    // Sync doctor order to Spring Boot backend
    const storeNum = parseInt(data.storeId.replace(/\D/g, ''), 10) || 1;
    const docUserIdNum = parseInt(data.doctorUser.id.replace(/\D/g, ''), 10) || 2;
    this.api.createDoctorOrder({
      storeId: storeNum,
      storeName: data.storeName,
      doctorUserId: docUserIdNum,
      doctorName: data.doctorUser.name,
      doctorRegNo: data.doctorUser.doctorRegNo || '',
      doctorSpecialty: data.doctorUser.doctorSpecialty || '',
      doctorPhone: data.doctorUser.phone || '',
      patientName: data.patient.patientName,
      patientAge: data.patient.patientAge,
      patientGender: data.patient.patientGender,
      patientPhone: data.patient.patientPhone,
      diagnosis: data.patient.diagnosis,
      patientVitalsBp: data.patient.vitalsBp || '',
      patientVitalsPulse: data.patient.vitalsPulse || '',
      patientVitalsWeight: data.patient.vitalsWeight || '',
      patientVitalsTemp: data.patient.vitalsTemp || '',
      patientVitalsSpo2: data.patient.vitalsSpo2 || '',
      prescriptionNotes: data.prescriptionNotes,
      deliveryAddress: data.deliveryAddress,
      paymentMethod: data.paymentMethod,
      items: data.items.map(i => ({
        medicineId: parseInt(i.medicineId.replace(/\D/g, ''), 10) || 1,
        medicineName: i.medicineName,
        genericName: i.genericName,
        packaging: i.packaging,
        quantity: i.quantity,
        unitPrice: i.unitPrice,
        dosage: i.dosage,
        timing: i.timing,
        durationDays: i.durationDays
      }))
    }).subscribe({
      next: (res) => {
        if (res?.orderNumber) {
          console.log('Doctor prescription registered in Spring Boot MySQL:', res.orderNumber);
        }
      },
      error: () => {}
    });

    return newOrder;
  }

  // 2. Customer Direct Order Creation
  createCustomerOrder(data: {
    storeId: string;
    storeName: string;
    customerUser: User;
    patient: OrderPatientInfo;
    items: PrescribedOrderItem[];
    deliveryAddress: string;
    prescriptionPhotoUrl?: string;
    paymentMethod: OrderPaymentMethod;
  }): StoreOrder {
    const totalAmount = data.items.reduce((sum, i) => sum + i.total, 0);
    const now = new Date();
    const orderNumber = 'ORD-CUST-' + String(this.orders().length + 1).padStart(4, '0');

    const newOrder: StoreOrder = {
      id: 'ord-' + Date.now(),
      orderNumber,
      orderType: 'CUSTOMER_ORDER',
      storeId: data.storeId,
      storeName: data.storeName,
      placedBy: {
        userId: data.customerUser.id,
        userName: data.customerUser.name,
        userRole: 'CUSTOMER',
        userPhone: data.customerUser.phone
      },
      patient: data.patient,
      items: data.items,
      deliveryAddress: data.deliveryAddress,
      prescriptionPhotoUrl: data.prescriptionPhotoUrl,
      paymentMethod: data.paymentMethod,
      paymentStatus: data.paymentMethod === 'ONLINE_PAID' ? 'PAID' : 'PENDING_COLLECTION',
      orderStatus: 'NEW_RECEIVED',
      totalAmount,
      createdAt: now.toISOString(),
      auditTrail: [
        {
          timestamp: now.toISOString(),
          action: `Customer Order Placed (${data.paymentMethod === 'COD' ? 'Cash on Delivery' : 'Prepaid Online'})`,
          performedBy: data.customerUser.name
        }
      ]
    };

    const updated = [newOrder, ...this.orders()];
    this.saveOrders(updated);

    // Sync customer order to Spring Boot backend
    const storeNum = parseInt(data.storeId.replace(/\D/g, ''), 10) || 1;
    const custUserIdNum = parseInt(data.customerUser.id.replace(/\D/g, ''), 10) || 3;
    this.api.createCustomerOrder({
      storeId: storeNum,
      storeName: data.storeName,
      customerUserId: custUserIdNum,
      customerName: data.customerUser.name,
      customerPhone: data.customerUser.phone || '',
      patientName: data.patient.patientName,
      familyMemberName: data.patient.familyMemberName || '',
      patientAge: data.patient.patientAge,
      patientGender: data.patient.patientGender,
      patientPhone: data.patient.patientPhone,
      diagnosis: data.patient.diagnosis,
      deliveryAddress: data.deliveryAddress,
      prescriptionPhotoUrl: data.prescriptionPhotoUrl,
      paymentMethod: data.paymentMethod,
      items: data.items.map(i => ({
        medicineId: parseInt(i.medicineId.replace(/\D/g, ''), 10) || 1,
        medicineName: i.medicineName,
        genericName: i.genericName,
        packaging: i.packaging,
        quantity: i.quantity,
        unitPrice: i.unitPrice,
        dosage: i.dosage,
        timing: i.timing,
        durationDays: i.durationDays
      }))
    }).subscribe({
      next: (res) => {
        if (res?.orderNumber) {
          console.log('Customer order registered in Spring Boot MySQL:', res.orderNumber);
        }
      },
      error: () => {}
    });

    return newOrder;
  }

  // 3. Store Owner Order Lifecycle Management
  updateOrderStatus(orderId: string, newStatus: OrderStatus, performedByName: string, note?: string): void {
    const now = new Date().toISOString();
    const updated = this.orders().map(order => {
      if (order.id !== orderId) return order;

      const auditTrail = [
        ...order.auditTrail,
        {
          timestamp: now,
          action: `Status changed to ${newStatus}${note ? ': ' + note : ''}`,
          performedBy: performedByName
        }
      ];

      const patch: Partial<StoreOrder> = {
        orderStatus: newStatus,
        auditTrail
      };

      if (newStatus === 'PACKED') patch.packedAt = now;
      if (newStatus === 'OUT_FOR_DELIVERY') patch.dispatchedAt = now;
      if (newStatus === 'COMPLETED') {
        patch.completedAt = now;
        if (order.paymentMethod === 'COD') patch.paymentStatus = 'PAID';
      }

      return {
        ...order,
        ...patch
      };
    });

    this.saveOrders(updated);

    // Sync status change with Spring Boot backend
    const numId = parseInt(orderId.replace(/\D/g, ''), 10);
    if (numId) {
      this.api.updateOrderStatus(numId, {
        newStatus,
        performedByName,
        notes: note || ''
      }).subscribe({
        next: () => {},
        error: () => {}
      });
    }
  }

  cancelOrder(orderId: string, reason?: string, cancelledBy?: string): void {
    this.updateOrderStatus(
      orderId,
      'CANCELLED',
      cancelledBy || 'Customer Self-Service',
      reason || 'Order cancelled by customer before dispatch'
    );
  }
}

File: javamedi/src/main/java/com/medi/app/service/OrderService.java
package com.medi.app.service;

import com.medi.app.dto.OrderDtos;
import com.medi.app.entity.OrderAuditLog;
import com.medi.app.entity.OrderItem;
import com.medi.app.entity.StoreOrder;
import com.medi.app.entity.User;
import com.medi.app.repository.OrderAuditLogRepository;
import com.medi.app.repository.StoreOrderRepository;
import com.medi.app.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Service
@RequiredArgsConstructor
public class OrderService {

    private final StoreOrderRepository storeOrderRepository;
    private final UserRepository userRepository;

    public List<StoreOrder> getAllOrders() {
        return storeOrderRepository.findAllByOrderByCreatedAtDesc();
    }

    public List<StoreOrder> getOrdersByStore(Long storeId) {
        return storeOrderRepository.findByStoreIdOrderByCreatedAtDesc(storeId);
    }

    public List<StoreOrder> getOrdersByUser(Long userId) {
        return storeOrderRepository.findByPlacedByUserIdOrderByCreatedAtDesc(userId);
    }

    public StoreOrder getOrderById(Long id) {
        return storeOrderRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Order not found with ID: " + id));
    }

    @Transactional
    public StoreOrder createDoctorOrder(OrderDtos.CreateDoctorOrderRequest req) {
        LocalDateTime now = LocalDateTime.now();
        String orderNumber = "ORD-DOC-" + String.format("%04d", storeOrderRepository.count() + 1);

        User doctor = userRepository.findById(req.getDoctorUserId())
                .orElseThrow(() -> new RuntimeException("Doctor user not found with ID: " + req.getDoctorUserId()));

        double totalAmount = 0.0;
        List<OrderItem> orderItems = new ArrayList<>();

        if (req.getItems() != null) {
            for (OrderDtos.PrescribedItemDto itemDto : req.getItems()) {
                double lineTotal = itemDto.getTotal() != null ? itemDto.getTotal() : (itemDto.getUnitPrice() * itemDto.getQuantity());
                totalAmount += lineTotal;

                OrderItem item = OrderItem.builder()
                        .medicineId(itemDto.getMedicineId())
                        .medicineName(itemDto.getMedicineName())
                        .genericName(itemDto.getGenericName())
                        .packaging(itemDto.getPackaging())
                        .quantity(itemDto.getQuantity())
                        .unitPrice(itemDto.getUnitPrice())
                        .total(lineTotal)
                        .dosage(itemDto.getDosage())
                        .timing(itemDto.getTiming())
                        .durationDays(itemDto.getDurationDays())
                        .build();

                orderItems.add(item);
            }
        }

        String paymentMethod = req.getPaymentMethod() != null ? req.getPaymentMethod() : "COD";
        String paymentStatus = "ONLINE_PAID".equalsIgnoreCase(paymentMethod) ? "PAID" : "PENDING_COLLECTION";

        StoreOrder order = StoreOrder.builder()
                .orderNumber(orderNumber)
                .orderType("DOCTOR_PRESCRIPTION")
                .storeId(req.getStoreId())
                .storeName(req.getStoreName())
                .placedByUserId(doctor.getId())
                .placedByUserName(doctor.getName())
                .placedByUserRole("DOCTOR")
                .placedByUserPhone(doctor.getPhone())
                .doctorRegNo(doctor.getDoctorRegNo())
                .doctorSpecialty(doctor.getDoctorSpecialty())
                .patientName(req.getPatientName())
                .patientAge(req.getPatientAge())
                .patientGender(req.getPatientGender())
                .patientPhone(req.getPatientPhone())
                .diagnosis(req.getDiagnosis())
                .prescriptionNotes(req.getPrescriptionNotes())
                .deliveryAddress(req.getDeliveryAddress())
                .paymentMethod(paymentMethod)
                .paymentStatus(paymentStatus)
                .totalAmount(Math.round(totalAmount * 100.0) / 100.0)
                .orderStatus("NEW_RECEIVED")
                .createdAt(now)
                .build();

        for (OrderItem oi : orderItems) {
            oi.setStoreOrder(order);
        }
        order.setItems(orderItems);

        // Audit Trail Entry
        OrderAuditLog auditLog = OrderAuditLog.builder()
                .timestamp(now)
                .action("Prescription Order Dispatched by Doctor")
                .performedBy(doctor.getName() + " (MCI: " + doctor.getDoctorRegNo() + ")")
                .storeOrder(order)
                .build();
        order.getAuditTrail().add(auditLog);

        return storeOrderRepository.save(order);
    }

    @Transactional
    public StoreOrder createCustomerOrder(OrderDtos.CreateCustomerOrderRequest req) {
        LocalDateTime now = LocalDateTime.now();
        String orderNumber = "ORD-CUST-" + String.format("%04d", storeOrderRepository.count() + 1);

        User customer = userRepository.findById(req.getCustomerUserId())
                .orElseThrow(() -> new RuntimeException("Customer user not found with ID: " + req.getCustomerUserId()));

        double totalAmount = 0.0;
        List<OrderItem> orderItems = new ArrayList<>();

        if (req.getItems() != null) {
            for (OrderDtos.PrescribedItemDto itemDto : req.getItems()) {
                double lineTotal = itemDto.getTotal() != null ? itemDto.getTotal() : (itemDto.getUnitPrice() * itemDto.getQuantity());
                totalAmount += lineTotal;

                OrderItem item = OrderItem.builder()
                        .medicineId(itemDto.getMedicineId())
                        .medicineName(itemDto.getMedicineName())
                        .genericName(itemDto.getGenericName())
                        .packaging(itemDto.getPackaging())
                        .quantity(itemDto.getQuantity())
                        .unitPrice(itemDto.getUnitPrice())
                        .total(lineTotal)
                        .build();

                orderItems.add(item);
            }
        }

        String paymentMethod = req.getPaymentMethod() != null ? req.getPaymentMethod() : "COD";
        String paymentStatus = "ONLINE_PAID".equalsIgnoreCase(paymentMethod) ? "PAID" : "PENDING_COLLECTION";

        StoreOrder order = StoreOrder.builder()
                .orderNumber(orderNumber)
                .orderType("CUSTOMER_ORDER")
                .storeId(req.getStoreId())
                .storeName(req.getStoreName())
                .placedByUserId(customer.getId())
                .placedByUserName(customer.getName())
                .placedByUserRole("CUSTOMER")
                .placedByUserPhone(customer.getPhone())
                .patientName(req.getPatientName() != null ? req.getPatientName() : customer.getName())
                .patientPhone(req.getCustomerPhone() != null ? req.getCustomerPhone() : customer.getPhone())
                .deliveryAddress(req.getDeliveryAddress())
                .paymentMethod(paymentMethod)
                .paymentStatus(paymentStatus)
                .totalAmount(Math.round(totalAmount * 100.0) / 100.0)
                .orderStatus("NEW_RECEIVED")
                .createdAt(now)
                .build();

        for (OrderItem oi : orderItems) {
            oi.setStoreOrder(order);
        }
        order.setItems(orderItems);

        OrderAuditLog auditLog = OrderAuditLog.builder()
                .timestamp(now)
                .action("Customer Order Placed (" + ("COD".equalsIgnoreCase(paymentMethod) ? "Cash on Delivery" : "Prepaid Online") + ")")
                .performedBy(customer.getName())
                .storeOrder(order)
                .build();
        order.getAuditTrail().add(auditLog);

        return storeOrderRepository.save(order);
    }

    @Transactional
    public StoreOrder updateOrderStatus(Long orderId, OrderDtos.UpdateOrderStatusRequest req) {
        StoreOrder order = getOrderById(orderId);
        LocalDateTime now = LocalDateTime.now();

        order.setOrderStatus(req.getNewStatus());

        if ("PACKED".equalsIgnoreCase(req.getNewStatus())) {
            order.setPackedAt(now);
        } else if ("OUT_FOR_DELIVERY".equalsIgnoreCase(req.getNewStatus())) {
            order.setDispatchedAt(now);
        } else if ("COMPLETED".equalsIgnoreCase(req.getNewStatus())) {
            order.setCompletedAt(now);
            if ("COD".equalsIgnoreCase(order.getPaymentMethod())) {
                order.setPaymentStatus("PAID");
            }
        }

        OrderAuditLog auditLog = OrderAuditLog.builder()
                .timestamp(now)
                .action("Order status updated to " + req.getNewStatus() + (req.getNotes() != null ? ": " + req.getNotes() : ""))
                .performedBy(req.getPerformedBy() != null ? req.getPerformedBy() : "Store Pharmacist")
                .notes(req.getNotes())
                .storeOrder(order)
                .build();

        order.getAuditTrail().add(auditLog);

        return storeOrderRepository.save(order);
    }
}
